AI Rerouting: Add OpenAI provider

shijithkjayan opened the pull request:
## Summary
 Target issue is [#4335](https://github.com/glific/glific/issues/4335)

Adds OpenAI provider to allow NGOs to bring their own OpenAI keys.

## ToDo
- [x] On creating credentials, send the keys to Kaapi.
- [ ] On updating credentials, send the keys to Kaapi.
- [ ] Use transaction for creating credentials. In case of errors, if we dont rollback we will run in to a problem where the credentials is present in Glific, but if missing in Kaapi resubmiting the form will fail as we will still be calling the update API.
- [ ] Write unit tests
- [ ] Delete the old `open_ai` provider


### Test Cases
1. Create a new credential with valid data
2. Create a new credential with invalid data
3. Update an existing credential with valid data
4. Update an existing credential with invalid data
5. Delete an existing credential in Kaapi (using their API) and then create a new one with valid data
6. Delete an existing credential in Kaapi (using their API)  and then create a new one with invalid data

PS: To test, run `mix phil.columns.seed` and then go to the settings page, where you can find a new provider called `OpenAI`.

AmishaBisht committed:
Add AI platform as a provider (#4308)

* feat: add AI platform as a provider

* refacto: remove unnecessary alias

* refactor: added a new migration file for ai platform provider

* refactor: adding the down fn in migration

* refactor: added kaapi insted of ai platform

AmishaBisht committed:
Add feature flag for ai-platform (#4309)

* feat: add feature flag for ai-platform

* test: test case for ai platform feature flag

* test: pass the test case

AmishaBisht committed:
Implement AI Platform Webhook Logic (#4311)

* feat: add re-routing to AI platform for filesearch gpt node

* feat: add re-routing to AI platform for filesearch gpt node

* build: bumped version to 7.4.0 (#4310)

* feat: wait for the ai platform to make the call to open ai without adding wait for result node

* refactor: remove unnecessary files

* refactor: remove unnecessary files

* refactor: add the updated webhook logs for ai response

* refactor: cleanup

* refactor: cleanup

* feat: get the kaapi API key from provider

* feat: add the test cases for flow_resume_controller

* feat: add the test cases for call a webhook

* test: add the test cases for call a webhook

* test: add more test cases

* refactor: store the result inside result variable

* refactor: fix gettext error

* refactor: update the feature flag name from ai platform to kaapi

* refactor: update the feature flag name from ai platform to kaapi

* refactor: update comments

* refactor: handle kaapi inactive case

* refactor: send error in appsignal

* refactor: send error in appsignal

* refactor: send error in appsignal

* refactor: add endpoint in the backend

* refactor: add endpoint in the backend

* refactor: add thread_id in kaapi response

* refactor: resolved reviews

* refactor: resolved reviews

* refactor: add tesla middleware

* refactor: fix the success response from kaapi

* refactor: fix the flow_resume_controller test cases

* refactor: fix the webhook_log_id format

* refactor: fix credo error

* refactor: handle the wait for result case in flow resume controller

* refactor: handle the wait for result case in flow resume controller

---------

Co-authored-by: Anandu Pavanan <pavanan.anandu63@gmail.com>

AmishaBisht committed:
Merge branch 'master' into feat/ai_platform_re-routing

madclaws committed:
fix: testing ffmpeg auto install

madclaws committed:
Merge branch 'master' into feat/ai_platform_re-routing

madclaws committed:
build: added this branch to deploy to staging

madclaws committed:
build: testing with cd to staging with label staging on PR

madclaws committed:
fix: wrong yml syntax

madclaws committed:
Merge branch 'master' into feat/ai_platform_re-routing

madclaws committed:
Merge branch 'master' into feat/ai_platform_re-routing

madclaws committed:
Merge branch 'master' into feat/ai_platform_re-routing

madclaws committed:
Merge branch 'master' into feat/ai_platform_re-routing

madclaws committed:
Merge branch 'master' into feat/ai_platform_re-routing

madclaws committed:
Merge branch 'master' into feat/ai_platform_re-routing

madclaws committed:
Merge branch 'master' into feat/ai_platform_re-routing

AmishaBisht committed:
Add kaapi API Keys in Providers Config per Org (#4348)

* feat: add function to onboard the orgs in kaapi and upadte their api keys in glific

* refactor: get the data fron db only insted of csv

* refactor: get the data fron db only insted of csv

* refactor: get the data fron db only insted of csv

* refactor: fix credo

* refactor: updated the module

* test: added the test case

* refactor: updated the config

* refactor: rename the kaapi module

* refactor: updated the action module for fetch kaapi function

* refactor: updated the migration module

* refactor: :recycle: Refactor Kaapi module structure and API client

- Move Kaapi credentials logic to dedicated Kaapi module - Update config
namespace to Kaapi.ApiClient module - Improve Tesla client configuration
with proper middleware - Extract configuration and headers into private
functions - Rename kaapi_keys_migration.ex to migration.ex for clarity

Co-authored-by: Amisha Bisht <amishabisht33@gmail.com>

* refactor: :recycle: Simplify Kaapi onboarding error handling

Remove complex error wrapping and async stream handling, update alias to
use ThirdParty.Kaapi directly, and rename upsert function to insert for
clarity.

Co-authored-by: Amisha Bisht <amishabisht33@gmail.com>

* test: :white_check_mark: Move Kaapi test to third_party directory and
add error handling tests

* refactor: :recycle: Refactor Kaapi onboarding to handle credential
creation

Move credential creation logic from Migration module to Kaapi module for
reusability and add proper error handling with logging to the onboard
function.

Co-authored-by: Amisha Bisht <amishabisht33@gmail.com>

* refactor: updated the middleware

* refactor: updated the middleware

* refactor: :recycle: Refactor onboard/2 to use pipe operator

* refactor: :recycle: Add global JSON middleware to Kaapi API client

* refactor: removed Tesla.post from onboard_to_kaapi

---------

Co-authored-by: Shijith K <shijith@Shijiths-T4D-MacBook-Pro.local>
Co-authored-by: Shijith Karumathil <shijithjayan@gmail.com>
Co-authored-by: Akansha Sakhre <asakhre2002@gmail.com>

madclaws committed:
Merge branch 'master' into feat/ai_platform_re-routing

akanshaaa19 committed:
AI Rerouting: sync existing assistants to kaapi (#4359)

* implement Kaapi provider with assistant ingestion functionality

* refactor Kaapi provider to improve assistant ingestion process and error handling

* refactor Kaapi provider to streamline assistant synchronization and improve documentation

* refactor: moved kaapi to third_party module

* made assistant fetching run in parallel for each org

* refactor: update organization filtering and default instructions in Kaapi ingest module

* refactor: enhance logging and error handling in Kaapi assistant synchronization

* refactor: improve organization fetching logic and encapsulate sync function

* refactor: remove variable for kaapi endpoint

* refactor: update module documentation for clarity on assistant synchronization

* refactor: simplify error handling in assistant fetching logic

* refactor: update kaapi endpoint retrieval method and fix fetch_assistants typo

* refactor: :recycle: Simplify process_orgs_concurrently with function
capture

- Out of scope of this ticket, but since its a simple refactor, added it
here.

* refactor: :recycle: Add ingest_ai_assistant function to Kaapi module

Refactors API calls to use a new wrapper function instead of calling
ApiClient directly, improving module encapsulation.

* fix: :bug: Fix error handling in KAAPI API response parser

Replace undefined variable `msg` with descriptive error message

* docs: :memo: Add documentation for Kaapi onboarding and ingestion
functions

* refactor: :recycle: Fix error handling and logging in Kaapi AI assistant
ingestion

Move logging from ApiClient to the main ingest function and improve
error response structure with status and message details.

Having logs both in the API client and in the context module is
redundant, so moved logging to the context module

* refactor: :recycle: Improve KAAPI error handling and simplify assistants
sync

Normalize error tuple handling, simplify return values by removing
org_id wrapper, fix query to return assistants list rather than tuple,
and consolidate instruction validation logic.

* fix: :bug: Use correct config path to get Kaapi endpoint configuration
and add alias

Add ApiClient alias and replace hardcoded endpoint concatenation with
Tesla BaseUrl middleware for cleaner configuration handling.

* refactor: :recycle: minor fixes

* ran mix format

* minor refactoring

* refactor KAAPI onboarding with success messages and increase concurrency

* fix: correct function name and update type specifications for ingesting AI assistants

* feat: Add OpenAI API key to Kaapi onboarding params

- Since our OpenAI keys are used for all NGOs, and the keys are required
in Kaapi when AI assistants are migrated as they validate the Assistants
with OpenAI, we are sending our OpenAI API keys for each org when they
are migrated/onboarded to Kaapi to allow the AI assistants sync to
happen.

* fix: Include OpenAI API key in Kaapi onboarding payload
fix: Update onboarding function to correctly handle API key response

* fix: Remove error logging for organization sync failure in KAAPI ingestion

* refactor: :recycle: Make openai_api_key optional and improve error
handling in Kaapi client

- Add conditional inclusion of openai_api_key in request payload -
Return detailed error response with status and body instead of generic
messages - Update tests to match new error response format

* chore: :globe_with_meridians: Update gettext line references for webhook
errors

* fic: :pencil2: Fix incorrect module name in Kaapi action test

- Still don't think the module name is correct, shuold be fixed later

* ci: :rotating_light: Update webhook log specs and add missing type specs
to fix doctor failures

Consolidates duplicate type specs for update_log/2 and adds missing spec
for webhook_and_wait/3 and ingest_ai_assistant/2. Updates return type to
include error case for better error handling.

* tests: :white_check_mark: Fix Kaapi action tests

- Remove hardcoded URL from Tesla mock patterns - Add proper assertions
for Action.execute return values - Replace database queries with direct
return value testing as the created messges are temporary and hence
never created and stored in database - Fix assertion case mismatch in
failure message test

* ci: :rotating_light: Update alias imports and fix Kaapi function type
specs to fix dialyzer errors

* ci: :rotating_light: Alphabetize imports in Kaapi.Ingest module

* ci: :rotating_light: Corrected wrong spec

* refactor: :recycle: Simplify result counting and update documentation for organization sync

* refactor: :recycle: Simplify organization assistant sync logic and improve logging

* refactor: :recycle: Simplify process_org_record function definition

* refactor: spec fix

---------

Co-authored-by: Shijith Karumathil <shijithjayan@gmail.com>
Co-authored-by: madclaws <pavanan.anandu63@gmail.com>

akanshaaa19 committed:
Refactor calculate_summary_stats to improve success count tracking for assistants

AmishaBisht committed:
AI re-routing: Use Kaapi APIs for Assistant Creation (#4332)

* refactor: changed the create assistant endpoint to kaapi

* refactoring: updated the endpoint for other open ai assistant operations

* refactoring: kaapi response handling

* test: add the test cases for assistant

* test: passed the test case

* refactor: update the import_assistant function

* refactor: fix the retrieve_assistant api

* refactor: update the changes in two kaapi calls

* refactor: update the update assistant api

* refactor: update the payload of create assistant in kaapi and update module

* refactor: update the dev file

* refactor: update the api client module and added the test cases

* refactor: :recycle: Add Kaapi assistant creation after OpenAI assistant

Moves Kaapi assistant creation to the main Filesearch module and adds
proper error handling with AppSignal reporting. Fixes the API client to
pass parameters correctly to the Kaapi service.

* typo: :pencil2: Fixed typo in error message

* refactor: :recycle: Refactor Kaapi API response parsing and fix type
specs

- Rename parse_onboard_response to parse_kaapi_response for reuse - Fix
organization_id parameter type in create_assistant functions - Update
function calls to use renamed response parser

* refactor: Refactor AI Assistants update to not fail even if Kaapi API
fails

- Moved Kaapi API call from Filesearch.ApiClient to Filesearch context
module. - Cleaned up Kaapi.ApiClient to use one generic parser function
for all Kaapi API responses. - Moved Kaapi API payload setup to Kaapi
context module

* refactor: :recycle: Extract assistant_id parameter and remove unused
code

Extracts assistant_id from params map in function signature, removes
duplicate API call line, and cleans up unused module imports.

* docs: :memo: Update KAAPI error log messages to use generic API
terminology

* feat: :sparkles: Add assistant deletion in Kaapi when removing
filesearch assistant

* docs: :memo: Add documentation for Kaapi module functions

* fix: :prencil2: Fix variable reference in Kaapi error message

Use organization_id instead of params.organization_id in the error
message string interpolation.

* refactor: :recycle: Simplify Kaapi API response parsing

Return full response body instead of extracting api_key field, making
the parser more flexible for different response formats.

* test: :white_check_mark: Added new Kaapi API tests and used describe
blocks

Group related tests into describe blocks, fix test descriptions, add
delete_assistant tests, and replace credential setup with module
constant for API key.

* Update assistant instructions and temperature handling in Kaapi and Filesearch API clients

* refactor: :recycle: Simplify Kaapi API error response handling

Remove conditional logic for different error body formats and always
return the raw response body as the error value.

* Add flow_id and contact_id to webhook log fields

* style: :art: Formatted code

* refactor: :recycle: For the nth time, simplified Kaapi API error
handling and responses cause I could not decide what approach to take

Remove custom error logging and message extraction logic as logging is
already done in Kaapi.ex. - Return structured error responses with
status and body instead of plain strings to give callers more context
for error handling.

* refactor: :recycle: Refactor ingest_ai_assistants to use client helper

- Remove `use Tesla` and simplify the function by using the existing
client/1 helper instead of building middleware inline.

* fix: dialyzer warning

---------

Co-authored-by: Shijith Karumathil <shijithjayan@gmail.com>
Co-authored-by: Akansha Sakhre <asakhre2002@gmail.com>
Co-authored-by: madclaws <pavanan.anandu63@gmail.com>

shijithkjayan committed:
Merge branch 'master' into feat/ai_platform_re-routing

shijithkjayan committed:
feat: :sparkles: Add OpenAI provider seed and test

madclaws commented:
`On creating credentials, send the keys to Kaapi.`

Are we doing this in this PR? @shijithkjayan

shijithkjayan commented:
> > On creating credentials, send the keys to Kaapi.

> Are we doing this in this PR? @shijithkjayan

Yes the Kaapi API is now ready, I will implement this today @madclaws

shijithkjayan committed:
refactor: :recycle: Simplify alias declarations in OpenAI provider seed

shijithkjayan committed:
feat: :sparkles: Create OpenAI credential in Glific and Kaapi

- Extract credential creation logic into private functions. - Add
validation callback for OpenAI credentials through Kaapi service. - Add
new disable_credential overload and create_credential API method.

shijithkjayan committed:
refactor: :recycle: Refactor function call and fix description
punctuation

- Use local function call instead of module-qualified call - Add missing
period to OpenAI provider description

shijithkjayan committed:
fix: :bug: Fix org ID source and update Kaapi credential function

- Use credential.organization_id - Rename create_credentials to
create_credential for consistency - Add trailing slash to credentials
API endpoint to avoid FastAPI errors

shijithkjayan committed:
feat: :sparkles: Add OpenAI credential update support via Kaapi

- Implements credential update callback for OpenAI provider that
validates API keys through Kaapi platform and disables credentials on
failure. - Includes new Kaapi API client method for PATCH credential
updates.

shijithkjayan committed:
refactor: :recycle: Fix OpenAI credential handling and error message

- Use credential.is_active instead of hardcoded true - Update error
message format for consistency

shijithkjayan committed:
fix: :bug: Fix variable reference in Kaapi error message

shijithkjayan committed:
fix: :bug: Fix OpenAI credential handling in partner callbacks

- Change credential.keys to credential.secrets for API key access - Add
provider field to update params

shijithkjayan committed:
feat: :sparkles: Add a catch all function clause for
credential_insert_callback/2 avoid errors when creating credentials
other than OpenAI
